### PR TITLE
allow to omit peerId and seqno

### PR DESCRIPTION
**Pull request by sinkingsugar**
related to https://github.com/ethereum/eth2.0-specs/issues/1981

### CODE REVIEW

File: libp2p/protocols/pubsub/gossipsub.nim [hunk 1]
```diff
@@ -1031,7 +1031,7 @@ method rpcHandler*(g: GossipSub,
 
     g.mcache.put(msgId, msg)
 
-    if g.verifySignature and not msg.verify(peer.peerId):
+    if not g.anonymize and g.verifySignature and not msg.verify(peer.peerId):
```
Comment (arnetheduck) [inline libp2p/protocols/pubsub/gossipsub.nim:1034]: there are a few more combinations to cover here:

* when the message has a signature, it must be validated - we should never propagate a message with an invalid signature - that includes discarding messages with empty bytes and other signatures / keys that can't be parsed
* when the message has a seqno, it must be valid as well (ie exactly 8 bytes long)
* if `verifySignature` is turned on, the message _must_ have a signature (and implicitly, it must be valid)
* if `anonymize` is turned on, we could either be strict (don't allow messages with signatures) or lax (allow messages with signatures) - for now, the latter is enough - we can add the strict checking at some other time.

Comment (sinkingsugar) [reply libp2p/protocols/pubsub/gossipsub.nim:1034]: Alright makes sense.

File: libp2p/protocols/pubsub/gossipsub.nim [hunk 2]
```diff
@@ -1031,11 +1031,21 @@ method rpcHandler*(g: GossipSub,
 
     g.mcache.put(msgId, msg)
 
-    if g.verifySignature and not msg.verify(peer.peerId):
+    if (msg.signature.len > 0 or g.verifySignature) and not msg.verify(peer.peerId):
```
Comment (arnetheduck) [inline libp2p/protocols/pubsub/gossipsub.nim:1034]: `peer.peerId` is not really needed for verify, not sure why it's there..?

**Review by arnetheduck — APPROVED** [review]

File: libp2p/protocols/pubsub/gossipsub.nim (hunk 2 above)
Comment (sinkingsugar) [reply libp2p/protocols/pubsub/gossipsub.nim:1034]: It's indeed not used

**Comment by dryajov** [comment]
Seems like this is changing again as per https://github.com/libp2p/specs/pull/294#issuecomment-697022095. We should probably hold off on it for the time being, or perhaps adopt a lax approach?

**Comment by arnetheduck** [comment]
yes, the lax mode is an intermediary step that is useful now

**Comment by sinkingsugar** [comment]
This PR was not complete, I said I needed a couple more hours on it.
The message cache won't work as the IDs are not generated properly.
We use the same style as go btw 
https://github.com/libp2p/go-libp2p-pubsub/blob/f7f33e10cc18b4a20542d0208aeaccb91ad64f99/pubsub.go#L965
But this doesn't work.

**Comment by sinkingsugar** [comment]
Since it's merged I will continue next week. As there is something weird going on with floodsub too.

**Comment by stefantalpalaru** [comment]
> This PR was not complete

Add "[WIP]" at the start of the title, while your PR is still "work in progress".

**Comment by arnetheduck** [comment]
right, forgot about that - need this for nbc to do something more reasnable soon - it's not affected at this point so.,

**Pull request merged**